core/src/queue.rs

```diff
@@ -192,15 +204,45 @@ impl Queue {
                 // When transactions are submitted quickly it can be reached.
                 Entry::Vacant(_) => continue,
             };
-            if self.check_tx(entry.get()).is_err() {
+            if self.check_tx(entry.get(), wsv).is_err() {
                 entry.remove_entry();
                 continue;
             }
 
             seen.push(hash);
             if *entry
                 .get()
-                .check_signature_condition(&self.wsv)
+                .check_signature_condition(wsv)
+                .expect("Checked in `check_tx` just above")
+            {
+                return Some(entry.get().clone());
+            }
+        }
+    }
+
+    /// Pops a single transaction.
+    #[allow(
+        clippy::expect_used,
+        clippy::unwrap_in_result,
+        clippy::cognitive_complexity
+    )]
+    pub fn pop_without_seen(&self, wsv: &WorldStateView) -> Option<VersionedAcceptedTransaction> {
```

Comment:
This is minor, but it feels like just adding a `bool` flag to normal `pop()` would remove the need for this function.

Reply:
I actually prefer it this way. `pop_without_seen` tells me volumes: 
1. There's a function `pop` whose default behaviour has to do with `seen`. 
2. Using this function is unambiguously not going to record things. Compared to that it's ambiguous which: `pop(true)` or `pop(false)` records txn. 
3. You see this as an option in intellisense (which I don't use, but most of you guys do). 
4. You don't have to change all instances of `pop` into `pop(false)`. 

This approach is an anti-pattern for most Python code I've seen. Thankfully we're not writing Python.